extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -353,97 +320,86 @@ export abstract class AzureAuth implements vscode.Disposable {
+		if (!tenant.id || !resource.id) {
+			Logger.pii('Tenant ID or resource ID was undefined', tenant, resource);
+			throw new AzureAuthError(msg, 'Adding account to cache failed', undefined);
 		}
-		const allSubs: Subscription[] = [];
-		const tokens = await this.getSecurityToken(account, azdata.AzureResource.ResourceManagement);
-		if (!tokens) {
-			Logger.log('There were no resource management tokens to retrieve subscriptions from. Account is stale.');
-			account.isStale = true;
+		try {
+			await this.tokenCache.saveCredential(`${accountKey.accountId}_access_${resource.id}_${tenant.id}`, JSON.stringify(accessToken));
+			await this.tokenCache.saveCredential(`${accountKey.accountId}_refresh_${resource.id}_${tenant.id}`, JSON.stringify(refreshToken));
+			this.memdb.set(`${accountKey.accountId}_${tenant.id}_${resource.id}`, expiresOn);
+		} catch (ex) {
+			Logger.error(ex);
+			throw new AzureAuthError(msg, 'Adding account to cache failed', ex);
 		}
+	}
 
-		for (const tenant of account.properties.tenants) {
-			const token = tokens[tenant.id];
-			const subscriptionUri = url.resolve(this.metadata.settings.armResource.endpoint, 'subscriptions?api-version=2019-11-01');
-			try {
-				const subscriptionResponse = await this.makeGetRequest(token.token, subscriptionUri);
-				Logger.pii('getSubscriptions', subscriptionResponse.data);
-				const subscriptions: Subscription[] = subscriptionResponse.data.value.map((subscriptionInfo: SubscriptionResponse) => {
-					return {
-						id: subscriptionInfo.subscriptionId,
-						displayName: subscriptionInfo.displayName,
-						tenantId: subscriptionInfo.tenantId
-					} as Subscription;
-				});
-				allSubs.push(...subscriptions);
-			} catch (ex) {
-				Logger.error(ex);
-				throw new Error('Error retrieving subscription information');
-			}
+	private async getSavedToken(tenant: Tenant, resource: Resource, accountKey: azdata.AccountKey): Promise<{ accessToken: AccessToken, refreshToken: RefreshToken, expiresOn: string }> {
+		const getMsg = localize('azure.cacheErrorGet', "Error when getting your account from the cache");
+		const parseMsg = localize('azure.cacheErrorParse', "Error when parsing your account from the cache");
+
+		if (!tenant.id || !resource.id) {
+			Logger.pii('Tenant ID or resource ID was undefined', tenant, resource);
+			throw new AzureAuthError(getMsg, 'Getting account from cache failed', undefined);
 		}
-		return allSubs;
-	}
 
-	protected async getToken(postData: { [key: string]: string }, tenant = this.commonTenant, resourceId: string = '', resourceEndpoint: string = ''): Promise<TokenRefreshResponse | undefined> {
+		let accessTokenString: string;
+		let refreshTokenString: string;
+		let expiresOn: string;
 		try {
-			let refreshResponse: TokenRefreshResponse;
+			accessTokenString = await this.tokenCache.getCredential(`${accountKey.accountId}_access_${resource.id}_${tenant.id}`);
```

Comment:
Make helper functions for these IDs so we don't risk typos breaking things

Reply:
I'm going to cover this with a test case anyway - don't want to add too much indirection - gets harder following the code.